### DISCUSSION TITLE
Schema design // How to handle polymorphism?

### DISCUSSION THREAD

**Discussion by evandjohnston** [prisma/prisma#3668] [Q&A] [answered]
### Question

Does anyone have examples of handling polymorphism in Prisma? I know [interfaces and unions aren’t supported](https://github.com/prisma/prisma/issues/2505), but I’m looking for repos where people have figured out their own solution.

### Goal & background

I'm trying to build a Typescript + GraphQL API server that mostly handles read requests on a strongly polymorphic dataset, as well as the usual user account management stuff.

I'm inexperienced with GraphQL and database design, so if it seems like I'm missing something obvious then I probably am.

### Example

Let's say our dataset contains `Animals`, with sub-types `Mammals`, `Reptiles`, `Birds`, etc., and sub-sub-types `Canines`, `Felids`, `Primates`, etc.

A user might request the list `(Mammals where hasWhiskers == true AND weight < 10 kg) AND  (Birds where weight > 2 kg)`. In actual usage, there will probably be several more conditions which may be specific to sub-sub-types, but I think this example is complex enough to illustrate the general idea.

### Desired traits of a solution

- Data typing is easy to validate
- Data integrity doesn't have to be guaranteed at the DB level, but mutation requests must be validated before persisting to DB
- Integration with GraphQL's interface type
- GraphQL queries are flatter as opposed to more deeply nested

#### 1. A single table with a field specifying the type, and additional nullable fields for every possible attribute.

- The only benefit here is that queries would be flat... but type validation would be a mess, validating mutation requests would be a mess, and GraphQL interface type integration would be a mess (or maybe impossible?).

#### 2. Each category has its own table, separate from its parents and children. Each parent category has a type column and a nullable reference column for each of its possible children.

- In this case, validating typing and mutations isn't too bad, but we end up with a deeply nested structure that is not well-adapted for use with GraphQL's features for addressing polymorphism.

### Wrap-up

I know I can't be the first person to have this problem, so how have other people solved it? Links to repos strongly preferred. 😉

Thanks for reading.

**Comment by ryands17** [comment, marked as answer]
Hey @evandjohnston 👋 

Yes currently Prisma doesn't support interfaces and unions, but as per your query

> I'm trying to build a Typescript + GraphQL API server that mostly handles read requests on a strongly polymorphic dataset, as well as the usual user account management stuff.

> Let's say our dataset contains Animals, with sub-types Mammals, Reptiles, Birds, etc., and sub-sub-types Canines, Felids, Primates, etc.

You are building a GraphQL API. Why not make use of Interfaces and Unions available to you in GraphQL itself instead of creating those in the database model? The best example I would suggest is something like [this](https://atheros.ai/blog/graphql-interfaces-and-unions-how-to-design-graphql-schema). This will help you add the required interfaces in the GraphQL layer with the separations that you require.

The above approach can be obtained by using an SDL first approach where you define your own schema in a `schema.graphql` file or use something like [Nexus](https://nexusjs.org/). Nexus also has an [interface type](https://nexusjs.org/components-standalone/schema/api/api-interfaceType#interfacetype) which could help you achieve that layer in GraphQL.

Let me know if that helps!

**Reply by evandjohnston** [reply]
Thanks, looks like Nexus will do what I need. 🙏

**Comment by lewebsimple** [comment]
Maybe  I’m missing something but how would the polymorphic fields be stored in  the database alongside normal prisma models?

**Reply by ryands17** [reply]
The fields will be stored normally as you would without polymorphism as Prisma doesn't support it currently. The workaround is to use Polymorphism in your application logic. So for e.g. if you're using GraphQL, you can use `Unions` and `Interfaces` to perform Polymorphism.

**Reply by aisensiy** [reply]
> The fields will be stored normally as you would without polymorphism as Prisma doesn't support it currently. The workaround is to use Polymorphism in your application logic. So for e.g. if you're using GraphQL, you can use `Unions` and `Interfaces` to perform Polymorphism.

So that means I should just create a model `Animal` with a `type` field in it and there is no more specific type on prisma level any more?

**Comment by bjufre** [comment]
I think that the Elixir library Ecto has a good page explaining how to deal with Polymorphic relationships, and some issues with the approach in frameworks/ORMs like Laravel or Rails.
Might be worth the read:
https://hexdocs.pm/ecto/polymorphic-associations-with-many-to-many.html

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
